### ISSUE TITLE
Reltable: Keyref to Topic in Other Keyscope Resolves to Wrong Instance of Linked Topic

### ISSUE THREAD

**Issue by edLearns** [dita-ot/dita-ot#3565]
Creating this issue in response to [this comment by @raducoravu](https://github.com/dita-ot/dita-ot/issues/2393#issuecomment-678265587).

## Expected Behavior
I’ve got two submaps, both of which reference a unique concept and the topic _t_topic.dita_ with a key defined in a separate _keydefTopic.ditamap_. Both submaps also have a ditaval to filter the content of _t_topic.dita_. The navigational topicref in each submap is also assigned a unique key, as in `<topicref keys="uniqueKey" keyref="previouslyDefinedKey"/>`.  
The unique key is then used in a reltable (in map one) to create links between the concept of map one and the specific instance of _t_topic.dita_ from map two. The topicref for the specific instance from map two uses the keyscope and the unique key, as in `<topicref keyref="map2.uniqueKey"/>`.  
I expect the concept from map one to be linked to the map two-instance of _t_topic.dita_ and vice versa.

## Actual Behavior
DITA-OT 3.5.2: The concept from map one is linked to the map one-instance of _t_topic.dita_ instead of the expected instance from map two. I get no output from DITA-OT 3.4.1.  
I used [@jelovirt’s fix of mapreflmpl.xsl](https://github.com/dita-ot/dita-ot/blob/hotfix/3.5.4/src/main/plugins/org.dita.base/xsl/preprocess/maprefImpl.xsl) for both DOT-versions.

## Steps to Reproduce
I’m attaching a small sample project to reproduce the problem. I also included logfiles for the html5-transformation with DITA-OT 3.4.1 as well as with DITA-OT 3.5.2:  
[20200824_githubSample.zip](https://github.com/dita-ot/dita-ot/files/5122703/20200824_githubSample.zip)

## Environment
* DITA-OT version: 3.4.1 and 3.5.2
* Operating system and version:
  _Windows_
* How did you run DITA-OT?
  _`dita` command_
* Transformation type:
  _HTML5_

**Comment by raducoravu** [comment]
This looks like a general problem in resolving key references in sibling key scopes:
http://docs.oasis-open.org/dita/dita/v1.3/errata02/os/complete/part3-all-inclusive/archSpec/base/using-keys-to-address-keys-across-scopes.html#concept_art_5ps_sr
If I open "c_aboutMap1.dita" and add an xref in it:

    <xref keyref="map2.m2-t_topic"/>

publish to HTML5 the reference will not be resolved to the proper topic instance in the "map2" sibling key scope.

**Comment by raducoravu** [comment]
So if we have two sibling key scopes "map1" and "map2", and I am inside "map1" according to the specs I can refer to keys defined in "map2" like "map2.m2-t_topic". This possibly works in a simpler context, but the current project also uses branch filtering.

**Comment by edLearns** [comment]
I somehow can’t see the general problem in resolving key references in sibling key scopes: If I use scope-qualified keyrefs, I should be fine, no? I did indeed use scope-qualified keys in my sample project and still got unexpected results.  

> This possibly works in a simpler context (…)  

I did not test the usage of scope-qualified keys without branch filtering. I naively expected things to just work. According to the specs, my expectation is justified. The DITA-OT I used to generate the output did not deliver the expected result. To me, this looks like another bug.

**Comment by raducoravu** [comment]
Yes, this is a bug. I do not disagree with this. Actually I was looking over this right now, but have not yet found the cause or a possible solution for it.

**Comment by raducoravu** [comment]
The method "org.dita.dost.module.KeyrefModule.walkMap(Element, KeyScope, List<ResolveTask>)" skips over all reltable elements because they are not topicrefs, so it does not recurse in the topicrefs inside the table cells. This does not seem to fix the problem, it's just an observation.

**Comment by raducoravu** [comment]
Adding some more details from my investigation:
The method "org.dita.dost.reader.KeyrefReader.resolveIntermediate(KeyScope, KeyDef, List<KeyDef>)" is at some point called for:

    <topicref keys="m2-t_topic" keyref="t_topic"/>

it's called in the key scope of "map1" and the "t_topic" is resolved to an href in that scope instead of being resolved in the key context of "map2".

**Comment by raducoravu** [comment]
I'm attaching a proposed fix in the java code, based on the latest 3.5.4 changes. It seems to work in this case and it did not seem to break any of our automated tests but I'm not 100% sure it's the best possible fix for this problem.
[KeyrefReader.java.zip](https://github.com/dita-ot/dita-ot/files/5161797/KeyrefReader.java.zip)

@edLearns if this issue is not fixed in the DITA OT you can drop me an email and I can offer you a late night build of the Oxygen publishing engine which is the DITA OT plus some commercial plugins and some patches.
